### PR TITLE
Backlog resilience

### PR DESCRIPTION
**Pull request by kettlebell**

### CODE REVIEW

File: spectrum-offchain/src/backlog.rs [hunk 1]
```diff
@@ -231,13 +231,20 @@ where
 impl<TOrd, TStore> BacklogService<TOrd, TStore>
 where
     TOrd: OnChainOrder + Weighted + Hash + Eq,
+    TOrd::TOrderId: Debug,
     TStore: BacklogStore<TOrd>,
 {
-    pub fn new<TOrd0: IsEqual<TOrd>>(store: TStore, conf: BacklogConfig) -> Self {
+    pub async fn new<TOrd0: IsEqual<TOrd>>(store: TStore, conf: BacklogConfig) -> Self {
```
Comment (kettlebell) [inline spectrum-offchain/src/backlog.rs:237]: The main change of this PR. The priority queues in `BacklogService` reside entirely in memory. If the bot ever requires a restart, the elements in the queues are lost.

So on program boot, we inspect the orders that reside in persistent storage and put them back into the `pending_pq` queue. This way we don't 'lose' orders if the bot shuts down.

File: ergo-chain-sync/src/client/node.rs [hunk 2]
```diff
@@ -27,46 +41,46 @@ impl ErgoNodeHttpClient {
 
 #[async_trait(?Send)]
 impl ErgoNetwork for ErgoNodeHttpClient {
-    async fn get_block_at(&self, height: u32) -> Option<FullBlock> {
+    async fn get_block_at(&self, height: u32) -> Result<FullBlock, Error> {
```
Comment (kettlebell) [inline ergo-chain-sync/src/client/node.rs:44]: It's quite helpful to know what errors are coming back from the ergo node.

File: spectrum-offchain/src/backlog.rs [hunk 3]
```diff
@@ -231,13 +231,20 @@ where
...
+        let mut pending_pq = PriorityQueue::new();
+        for ord in store.find_orders(|_| true).await {
```
Comment (oskin1) [inline spectrum-offchain/src/backlog.rs:239]: Why not just creat something like `get_all()` or `stream_all()`?

Comment (kettlebell) [reply spectrum-offchain/src/backlog.rs:239]: Actually, it's currently used now to remove unnecessary orders once it's confirmed that no stakers are left for a given epoch: https://github.com/spectrum-finance/spectrum-offchain-ergo/blob/73ac89189fd755f91c24445b816c2f42c57c01c5/spectrum-offchain-lm/src/scheduler/process.rs#L68-L79

File: spectrum-offchain-lm/src/bundle/rocksdb.rs [hunk 4]
```diff
@@ -156,8 +156,8 @@ impl BundleRepo for BundleRepoRocksDB {
             let tx = db.transaction();
             tx.put(state_key, state_bytes).unwrap();
             tx.put(index_key, state_id_bytes).unwrap();
-            tx.delete(prev_epoch_index_key).unwrap();
             tx.put(epoch_index_key, dummy_bytes).unwrap();
+            tx.delete(prev_epoch_index_key).unwrap();
```
Comment (oskin1) [inline spectrum-offchain-lm/src/bundle/rocksdb.rs:160]: Does this make any difference?

**Review by oskin1 — APPROVED** [review]

File: spectrum-offchain-lm/src/bundle/rocksdb.rs (hunk 4 above)
Comment (kettlebell) [reply spectrum-offchain-lm/src/bundle/rocksdb.rs:160]: No difference.

**Pull request merged**